### PR TITLE
Allow URI as data for vcard PHOTO

### LINKED ISSUE #13085: CardDAV\PhotoCache should support Binary
<h3>Technical details</h3>
<ul>
	<li>Remote Address: 172.24.0.1</li>
	<li>Request ID: Eyp0YLWK8j0XvcMPix5v</li>
	<li>Type: TypeError</li>
	<li>Code: 0</li>
	<li>Message: Argument 1 passed to OCA\DAV\CardDAV\PhotoCache::getType() must be an instance of Sabre\VObject\Property\Binary, instance of Sabre\VObject\Property\Uri given, called in /var/www/html/apps/dav/lib/CardDAV/PhotoCache.php on line 196</li>
	<li>File: /var/www/html/apps/dav/lib/CardDAV/PhotoCache.php</li>
	<li>Line: 243</li>
</ul>

<h3>Trace</h3>
<pre>#0 /var/www/html/apps/dav/lib/CardDAV/PhotoCache.php(196): OCA\DAV\CardDAV\PhotoCache-&gt;getType(Object(Sabre\VObject\Property\Uri))
#1 /var/www/html/apps/dav/lib/CardDAV/PhotoCache.php(89): OCA\DAV\CardDAV\PhotoCache-&gt;getPhoto(Object(Sabre\CardDAV\Card))
#2 /var/www/html/apps/dav/lib/CardDAV/PhotoCache.php(62): OCA\DAV\CardDAV\PhotoCache-&gt;init(Object(OC\Files\SimpleFS\SimpleFolder), Object(Sabre\CardDAV\Card))
#3 /var/www/html/apps/dav/lib/CardDAV/ImageExportPlugin.php(104): OCA\DAV\CardDAV\PhotoCache-&gt;get(&#039;1&#039;, &#039;3CCCE266-C4D9-4...&#039;, -1, Object(Sabre\CardDAV\Card))
#4 [internal function]: OCA\DAV\CardDAV\ImageExportPlugin-&gt;httpGet(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
#5 /var/www/html/3rdparty/sabre/event/lib/EventEmitterTrait.php(105): call_user_func_array(Array, Array)
#6 /var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php(479): Sabre\Event\EventEmitter-&gt;emit(&#039;method:GET&#039;, Array)
#7 /var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php(254): Sabre\DAV\Server-&gt;invokeMethod(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
#8 /var/www/html/apps/dav/lib/Server.php(298): Sabre\DAV\Server-&gt;exec()
#9 /var/www/html/apps/dav/appinfo/v2/remote.php(35): OCA\DAV\Server-&gt;exec()
#10 /var/www/html/remote.php(163): require_once(&#039;/var/www/html/a...&#039;)
#11 {main}</pre>

### PR DESCRIPTION
**Pull request by skjnldsv**
Fix #13085 

Failing vcard example: [6644643C-FF79-4044-82DC-5A2776EDD73D.txt](https://github.com/nextcloud/server/files/2685864/6644643C-FF79-4044-82DC-5A2776EDD73D.txt)

### CODE REVIEW

**Comment by nickvergessen** [comment]
Does this fix further issues from 
https://github.com/nextcloud/server/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3A%22feature%3A+dav%22+photo

at least #3366 and #4358 sound related?

**Comment by skjnldsv** [comment]
#4358: definitely! I'll investigate #3366! :)
Thanks @nickvergessen :hugs:

File: apps/dav/lib/CardDAV/PhotoCache.php [hunk 1]
```diff
@@ -237,10 +237,10 @@ private function readCard($cardData) {
 	}
 
 	/**
-	 * @param Binary $photo
+	 * @param Binary|Uri $photo
 	 * @return string
 	 */
-	private function getType(Binary $photo) {
+	private function getType(Property $photo) {
```
Comment (georgehrke) [inline apps/dav/lib/CardDAV/PhotoCache.php]: What will `getType` return in the case of your example vcard? just an empty string? 
Shouldn't it try to parse the `image/jpeg;base64` or is that handled elsewhere?

Comment (skjnldsv) [reply apps/dav/lib/CardDAV/PhotoCache.php]: Fixed!
I misunderstood how the file worked.
Now it's better :)

**Comment by skjnldsv** [comment]
Sorry about the push mess. I'm tired it seems.... :zzz:

**Comment by skjnldsv** [comment]
Failure unrelated! Please review! :)

**Review by georgehrke — APPROVED** [review]

**Comment by skjnldsv** [comment]
Bump! :)

File: apps/dav/lib/CardDAV/PhotoCache.php [hunk 2]
```diff
@@ -193,19 +199,23 @@ private function getPhoto(Card $node) {
 			}
 
 			$photo = $vObject->PHOTO;
-			$type = $this->getType($photo);
 
-			$val = $photo->getValue();
+			// handle data URI. e.g PHOTO;VALUE=URI:data:image/jpeg;base64,/9j/4AAQSkZJRgABAQE
 			if ($photo->getValueType() === 'URI') {
 				$parsed = \Sabre\URI\parse($val);
```
Comment (MorrisJobke) [inline apps/dav/lib/CardDAV/PhotoCache.php]: Isn't `$val` undefined here?

File: apps/dav/lib/CardDAV/PhotoCache.php [hunk 3]
```diff
@@ -193,19 +199,23 @@ private function getPhoto(Card $node) {
...
+
+				// only allow data://
 				if ($parsed['scheme'] !== 'data') {
 					return false;
 				}
 				if (substr_count($parsed['path'], ';') === 1) {
-					list($type,) = explode(';', $parsed['path']);
+					list($type) = explode(';', $parsed['path']);
 				}
 				$val = file_get_contents($val);
```
Comment (MorrisJobke) [inline apps/dav/lib/CardDAV/PhotoCache.php]: Isn't `$val` undefined here?

File: apps/dav/lib/CardDAV/PhotoCache.php [hunk 4]
```diff
@@ -214,16 +224,18 @@ private function getPhoto(Card $node) {
...
 
 			return [
 				'Content-Type' => $type,
-				'body' => $val
+				'body'         => $val
 			];
-		} catch(\Exception $ex) {
-
+		} catch (\Exception $e) {
+			$this->logger->logException($ex, [
```
Comment (MorrisJobke) [inline apps/dav/lib/CardDAV/PhotoCache.php]:
```suggestion
			$this->logger->logException($e, [
```

File: apps/dav/lib/CardDAV/PhotoCache.php (hunk 3 above)
Comment (skjnldsv) [reply apps/dav/lib/CardDAV/PhotoCache.php]: Somehow failed the previous commit yes :)

**Review by MorrisJobke — APPROVED** [review]
Code makes sense 👍

**Comment by MorrisJobke** [comment]
rebased and autosquashed

**Comment by MorrisJobke** [comment]
> The backport to stable14 failed. Please do this backport manually.

@skjnldsv Mind to do this?

**Comment by skjnldsv** [comment]
Backport in https://github.com/nextcloud/server/pull/13708

**Pull request merged**